### PR TITLE
Implement Confluence DLS

### LINKED ISSUE #1291: DLS: Confluence, Jira
Enhance the 2 mentioned connectors with document level security.

### Acceptance criteria
- [ ] discovery document with DLS for
    1. Confluence
    2. Jira
- [ ] add support for DLS as defined in the discovery
- [ ] functional tests performed
- [ ] documentation with several examples illustrating the syntax, how a basic use case works, limitations
- [ ] performance test/ reports against the last stable release branch of the connector framework and in Elastic Cloud

### LINKED ISSUE #1399: add DLS support for Confluence as defined in the discovery

### PR DESCRIPTION
**Pull request by abhishekjoshi-crest**
## Closes #1399 

Adds the document level security for Confluence connector.

#### Pre-Review Checklist
- [x] this PR has a meaningful title
- [x] this PR links to all relevant github issues that it fixes or partially addresses
- [x] if there is no GH issue, please create it. Each PR should have a link to an issue
- [x] this PR has a thorough description
- [x] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [x] Considered corresponding documentation changes
- [x] Contributed any configuration settings changes to the configuration reference

### CODE REVIEW

File: connectors/sources/confluence.py [hunk 1]
```diff
@@ -61,6 +64,31 @@
 CONFLUENCE_SERVER = "confluence_server"
 WILDCARD = "*"
 
+ACCESS_CONTROL = "_allow_access_control"
+
+
+def _prefix_identity(prefix, identity):
```
Comment (timgrein) [inline connectors/sources/confluence.py:70]: I've noted in the Jira PR, that we should move these functions to a common `access_control.py` module, but that'll be on my side 👍

File: connectors/sources/confluence.py [hunk 2]
```diff
@@ -312,8 +341,230 @@ def get_default_configuration(cls):
...
+                    "_id": <account_id>,
+                    "identity": {
+                        "account_id": <_prefixed_account_id>,
+                        "display_name": <_preffixed_account_name>
+                    },
+                    "created_at": <iso_utc_timestamp>,
+                    ACCESS_CONTROL: [<_prefixed_account_id>, <_prefixed_group_ids>, <_prefixed_role_keys>]
+                }
+        """
+        account_id = user.get("accountId")
```
Comment (timgrein) [inline connectors/sources/confluence.py:437]: This whole method looks very similar to the one in the Jira connector. Should we create a common `atlassian.py` module to be able to reuse this functionality?

File: connectors/sources/confluence.py [hunk 3]
```diff
@@ -312,8 +341,230 @@ def get_default_configuration(cls):
...
+        async for users in self.confluence_client.api_call(
+            url=os.path.join(self.configuration["confluence_url"], URLS[USER])
+        ):
+            yield await users.json()
+
+    async def fetch_user(self, user_url):
+        async for user in self.confluence_client.api_call(
+            url=f"{user_url}&{USER_QUERY}"
+        ):
+            yield await user.json()
```
Comment (timgrein) [inline connectors/sources/confluence.py:479]: suggestion: I would move these methods to the `JiraClient` and make the `api_call` method "private".

File: connectors/sources/confluence.py [hunk 4]
```diff
@@ -312,8 +341,230 @@ def get_default_configuration(cls):
...
+
+    def _is_active_atlassian_user(self, user_info):
+        user_url = user_info.get("self")
+        if not user_url:
+            self._logger.debug(
+                f"Skipping {user_info.get('displayName', 'user')} as profile URL is not present."
```
Comment (timgrein) [inline connectors/sources/confluence.py:485]: suggestion: I would introduce a variable for `user_info.get('displayName', 'user')` and reuse it throughout this method.

File: connectors/sources/confluence.py [hunk 5]
```diff
@@ -312,8 +341,230 @@ def get_default_configuration(cls):
...
+    def _extract_identities(self, response):
+        if not self._dls_enabled():
+            return set()
+
+        identities = set()
+        user_results = response.get("user", {}).get("results", [])
+        group_results = response.get("group", {}).get("results", [])
+
+        for item in user_results + group_results:
+            if item.get("type") == "known" and item.get("accountType") == "atlassian":
```
Comment (timgrein) [inline connectors/sources/confluence.py:561]: nit: `item.get("type")` could be a variable `item_type`.

File: connectors/sources/confluence.py [hunk 6]
```diff
@@ -588,20 +854,43 @@ async def _attachment_coro(self, document):
...
-    async def _page_blog_coro(self, api_query):
+    async def _page_blog_coro(self, api_query, target_type):
         """Coroutine to add pages/blogposts to Queue
 
         Args:
             api_query (str): API Query Parameters for fetching page/blogpost
+            target_type (str): Type of object to filter permission
         """
-        async for document, attachment_count in self.fetch_documents(api_query):
+        async for document, attachment_count, permissions, restrictions in self.fetch_documents(
```
Comment (timgrein) [inline connectors/sources/confluence.py:876]: The return type of `fetch_documents` seems to be pretty complex. Can we simplify it? Also what's the difference between `permissions` and `restrictions`?

**Review by timgrein — COMMENTED** [review]
Looks good from a functional perspective. Left some comments around class structure and readability.

File: connectors/sources/confluence.py (hunk 2 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/confluence.py:437]: Yes this was in my mind to move common code in `atlassian.py`(Already created this common module during Advanced Sync Rules implementation) once both DLS implementation will merge we will raise one incremental PR for the same 👍

File: connectors/sources/confluence.py (hunk 6 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/confluence.py:876]: Permission are related to space and Restriction are related to pages & blogpost (Implement side need to override space permission by pages & blogpost restriction). [Difference Reference](https://support.atlassian.com/confluence-cloud/docs/what-are-confluence-cloud-permissions-and-restrictions/).

File: connectors/sources/confluence.py (hunk 3 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/confluence.py:479]: Confluence & Jira indexing implementation includes only `api_call` & `paginated_call` in client so keeping identical through out connector for client-base approach

**Comment by abhishekjoshi-crest** [comment]
buildkite test this

File: connectors/sources/confluence.py (hunk 2 above)
Comment (timgrein) [reply connectors/sources/confluence.py:437]: Thanks! Can you create a corresponding issue so we can track it?

File: connectors/sources/confluence.py (hunk 3 above)
Comment (timgrein) [reply connectors/sources/confluence.py:479]: If they're identical we could have an `AtlassianClient` client inside the `atlassian.py` module, which we could reuse in the specific `JiraClient` and `ConfluenceClient` classes, right?

File: connectors/sources/confluence.py (hunk 6 above)
Comment (timgrein) [reply connectors/sources/confluence.py:874]: Could we add some comment explaining this difference in the code?

File: connectors/sources/confluence.py (hunk 2 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/confluence.py:437]: Yes @moxarth-elastic created it https://github.com/elastic/connectors-python/issues/1417

File: connectors/sources/confluence.py (hunk 6 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/confluence.py:874]: Sure!

File: connectors/sources/confluence.py (hunk 3 above)
Comment (abhishekjoshi-crest) [reply connectors/sources/confluence.py:479]: Both are identical we can have it in `AtlassianClient` 👍

**Review by timgrein — APPROVED** [review]
LGTM, thank you for addressing my comments 👍

**Pull request merged**